core/arch/riscv64/mangle.c

```diff
@@ -612,37 +612,19 @@ instr_t *
 mangle_rel_addr(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
                 instr_t *next_instr)
 {
-    opnd_t dst = instr_get_dst(instr, 0);
     app_pc tgt;
     ASSERT(instr_get_opcode(instr) == OP_auipc);
     ASSERT(instr_has_rel_addr_reference(instr));
     instr_get_rel_data_or_instr_target(instr, &tgt);
-    ASSERT(opnd_is_reg(dst));
-    ASSERT(opnd_is_rel_addr(instr_get_src(instr, 0)));
-    bool uses_reg_stolen = instr_uses_reg(instr, dr_reg_stolen),
-         uses_reg_tp = instr_uses_reg(instr, DR_REG_TP);
 
-    if (uses_reg_stolen || uses_reg_tp) {
-        dst = opnd_create_reg(DR_REG_A0);
-        PRE(ilist, next_instr,
-            instr_create_save_to_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
-    }
+    mangle_special_registers(dcontext, ilist, instr, next_instr);
 
-    insert_mov_immed_ptrsz(dcontext, (ptr_int_t)tgt, dst, ilist, next_instr, NULL, NULL);
+    opnd_t dst = instr_get_dst(instr, 0);
+    ASSERT(opnd_is_reg(dst));
+    ASSERT(opnd_is_rel_addr(instr_get_src(instr, 0)));
 
-    if (uses_reg_stolen) {
-        PRE(ilist, next_instr,
-            instr_create_save_to_tls(dcontext, DR_REG_A0, TLS_REG_STOLEN_SLOT));
-        PRE(ilist, next_instr,
-            instr_create_restore_from_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
-    } else if (uses_reg_tp) {
-        PRE(ilist, next_instr,
-            instr_create_save_to_tls(
-                dcontext, DR_REG_A0,
-                os_tls_offset(os_get_app_tls_base_offset(TLS_REG_LIB))));
-        PRE(ilist, next_instr,
-            instr_create_restore_from_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
-    }
+    insert_mov_immed_ptrsz(dcontext, (ptr_int_t)tgt, dst, ilist, instr_get_next(instr),
```

Comment:
Why instr_get_next(instr)?  Just `instr` seems more straightforward?

Reply:
Yes, `instr` is removed later and inserting the mov whether before or later doesn't makes a difference.